src/Features/Core/Portable/SplitOrMergeIfStatements/Nested/AbstractMergeNestedIfStatementsCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,199 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Utilities;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.SplitOrMergeIfStatements
+{
+    internal abstract class AbstractMergeNestedIfStatementsCodeRefactoringProvider
+        : AbstractMergeIfStatementsCodeRefactoringProvider
+    {
+        // Converts:
+        //    if (a)
+        //    {
+        //        if (b)
+        //            Console.WriteLine();
+        //    }
+        //
+        // To:
+        //    if (a && b)
+        //        Console.WriteLine();
+
+        protected sealed override CodeAction CreateCodeAction(Func<CancellationToken, Task<Document>> createChangedDocument, string ifKeywordText)
+            => new MyCodeAction(createChangedDocument, ifKeywordText);
+
+        protected sealed override async Task<bool> CanBeMergedAsync(
+            Document document, SyntaxNode ifStatement, CancellationToken cancellationToken)
+        {
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            var ifGenerator = document.GetLanguageService<IIfLikeStatementGenerator>();
+
+            return IsFirstStatementOfIfLikeStatement(syntaxFacts, ifGenerator, ifStatement, out var outerIfLikeStatement) &&
+                   await CanBeMergedWithOuterAsync(document, syntaxFacts, ifGenerator, outerIfLikeStatement, ifStatement, cancellationToken).ConfigureAwait(false);
+        }
+
+        protected sealed override SyntaxNode GetChangedRoot(Document document, SyntaxNode root, SyntaxNode ifStatement)
+        {
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            var ifGenerator = document.GetLanguageService<IIfLikeStatementGenerator>();
+            var generator = document.GetLanguageService<SyntaxGenerator>();
```

Comment:
consider tuples.

Reply:
I don't like the idea of introducing a new base class just for this one helper.